x-pack/plugins/event_log/server/es/cluster_client_adapter.ts

```diff
@@ -30,14 +42,71 @@ export interface QueryEventsBySavedObjectResult {
 export class ClusterClientAdapter {
   private readonly logger: Logger;
   private readonly clusterClientPromise: Promise<EsClusterClient>;
+  private readonly docBuffer$: Subject<Doc>;
+  private readonly docsBufferedSubscription: Subscription;
+  private readonly doneWriting: ReadySignal;
+  private readonly context: EsContext;
 
   constructor(opts: ConstructorOpts) {
     this.logger = opts.logger;
     this.clusterClientPromise = opts.clusterClientPromise;
+    this.context = opts.context;
+    this.docBuffer$ = new Subject<Doc>();
+    this.doneWriting = createReadySignal();
+
+    // buffer event log docs for time / buffer length, ignore empty
+    // buffers, then index the buffered docs
+    const docsBuffered$ = this.docBuffer$.pipe(
+      bufferTime(EVENT_BUFFER_TIME, null, EVENT_BUFFER_LENGTH),
+      filter((docs) => docs.length > 0),
+      switchMap(async (docs) => await this.indexDocuments(docs))
+    );
+
+    // kick everything off, when the stream closes, signal finished
+    const { doneWriting } = this;
+    this.docsBufferedSubscription = docsBuffered$.subscribe({
+      complete() {
+        doneWriting.signal();
+      },
+    });
+  }
+
+  // This will be called at plugin stop() time; the assumption is any plugins
+  // depending on the event_log will already be stopped, and so will not be
+  // writing more event docs.  So we shut down the docBuffer$ observable,
+  // and wait for the docsBufffered$ observable to complete
+  public async shutdown(): Promise<void> {
+    this.docBuffer$.complete();
+    await this.doneWriting.wait();
+    this.docsBufferedSubscription.unsubscribe();
+  }
+
+  public indexDocument(doc: Doc): void {
+    this.docBuffer$.next(doc);
   }
 
-  public async indexDocument(doc: unknown): Promise<void> {
-    await this.callEs<ReturnType<Client['index']>>('index', doc);
+  async indexDocuments(docs: Doc[]): Promise<void> {
+    // if es initialization failed, don't try to index
+    if (!(await this.context.waitTillReady())) {
+      return;
```

Comment:
Should this throw an error instead of just returning silently? 🤔 
Or at least log it?

Reply:
Originally, it did log.  However, I realized that it would log anytime the messages got flushed, and the entire pipeline will still run even if the initialization failed (indicated by `waitTillReady() resolving to `false`).  The entire pipeline still runs because the code path for logging an event from a client is now **completely** synchronous, replacing the old `setImmediate()` forking of the log writing, to the new `next()` on the observable - this is fabulous!  But also means we can't check `waitTillReady()` until we get to this point.  And we won't want to be spamming an error message here :-)

So, instead, I put one of these `waitTillReady()` calls with a single log message in the plugin code itself: https://github.com/elastic/kibana/blob/1a7134dfb5a5056448a8624d116530006782b6d0/x-pack/plugins/event_log/server/plugin.ts#L120-L125

I'm not convinced this is the best approach, perhaps `waitTillReady()` should do the logging itself, in it's error paths, but then that's really too specific - I'm kinda treating `waitTillReady()` as a boolean getter at this point, felt safer to not have it log specific outcomes itself as a side effect.

But clearly needs a comment, in both locations, it's certainly not completely obvious.